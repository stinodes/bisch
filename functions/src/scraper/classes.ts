import * as puppeteer from 'puppeteer'

export const scrapeClasses = async () => {
  const url = 'https://www.wowhead.com/wotlk/'
  const browser = await puppeteer.launch({
    headless: true,
  })
  const page = await browser.newPage()
  page.on('console', msg => console.log('PAGE LOG:', msg.text()))
  await page.goto(url)

  const el = await page.waitForSelector('#onetrust-consent-sdk')
  if (el) {
    await el.evaluate(el => el.remove())
  }

  const nav = await page.waitForSelector('div.header-nav-features')

  if (!nav) throw new Error('Could not correctly parse Wowhead.')

  const boundingBox = (await nav.boundingBox()) || { x: 0, y: 0 }

  page.mouse.move(boundingBox.x + 10, boundingBox.y + 10)

  await page.waitForSelector('div.menu div.menu-item > a')
  const specs = await page.$$eval('div.menu div.menu-item > a', items => {
    return items
      .map(el => {
        if (!el || el.classList.contains('menu-item-heading')) return null

        const color = getComputedStyle(el).color
        const icon = getComputedStyle(el).backgroundImage.split('"')[1]
        const name = el.textContent?.trim() || ''

        return {
          name,
          slug: name.toLowerCase().replace(' ', '-'),
          color,
          icon,
        }
      })
      .filter(Boolean)
  })
  await browser.close()

  return specs
}
