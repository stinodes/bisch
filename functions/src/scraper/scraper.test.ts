import { scrapeClasses } from './classes'

describe('Scraper', () => {
  test('fetches all specs and their metadata', async () => {
    const data = await scrapeClasses()

    expect(data).toBeInstanceOf(Array)
    expect(data.length).toBe(31)
    expect(data).toMatchSnapshot()
  })
})
