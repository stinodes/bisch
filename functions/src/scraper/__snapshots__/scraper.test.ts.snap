// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scraper fetches all specs and their metadata 1`] = `
[
  {
    "color": "rgb(196, 30, 58)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/spell_deathknight_bloodpresence.gif",
    "name": "Blood Death Knight",
    "slug": "blood-death knight",
  },
  {
    "color": "rgb(196, 30, 58)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/spell_deathknight_frostpresence.gif",
    "name": "Frost Death Knight",
    "slug": "frost-death knight",
  },
  {
    "color": "rgb(196, 30, 58)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/spell_deathknight_unholypresence.gif",
    "name": "Unholy Death Knight",
    "slug": "unholy-death knight",
  },
  {
    "color": "rgb(255, 124, 10)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/spell_nature_starfall.gif",
    "name": "Balance Druid",
    "slug": "balance-druid",
  },
  {
    "color": "rgb(255, 124, 10)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/ability_druid_catform.gif",
    "name": "Feral DPS",
    "slug": "feral-dps",
  },
  {
    "color": "rgb(255, 124, 10)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/ability_racial_bearform.gif",
    "name": "Feral Tank",
    "slug": "feral-tank",
  },
  {
    "color": "rgb(255, 124, 10)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/spell_nature_healingtouch.gif",
    "name": "Restoration Druid",
    "slug": "restoration-druid",
  },
  {
    "color": "rgb(170, 211, 114)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/ability_hunter_beasttaming.gif",
    "name": "Beast Mastery Hunter",
    "slug": "beast-mastery hunter",
  },
  {
    "color": "rgb(170, 211, 114)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/ability_marksmanship.gif",
    "name": "Marksmanship Hunter",
    "slug": "marksmanship-hunter",
  },
  {
    "color": "rgb(170, 211, 114)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/ability_hunter_swiftstrike.gif",
    "name": "Survival Hunter",
    "slug": "survival-hunter",
  },
  {
    "color": "rgb(63, 199, 235)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/spell_holy_magicalsentry.gif",
    "name": "Arcane Mage",
    "slug": "arcane-mage",
  },
  {
    "color": "rgb(63, 199, 235)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/spell_fire_firebolt02.gif",
    "name": "Fire Mage",
    "slug": "fire-mage",
  },
  {
    "color": "rgb(63, 199, 235)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/spell_frost_frostbolt02.gif",
    "name": "Frost Mage",
    "slug": "frost-mage",
  },
  {
    "color": "rgb(244, 140, 186)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/spell_holy_holybolt.gif",
    "name": "Holy Paladin",
    "slug": "holy-paladin",
  },
  {
    "color": "rgb(244, 140, 186)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/spell_holy_devotionaura.gif",
    "name": "Protection Paladin",
    "slug": "protection-paladin",
  },
  {
    "color": "rgb(244, 140, 186)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/spell_holy_auraoflight.gif",
    "name": "Retribution Paladin",
    "slug": "retribution-paladin",
  },
  {
    "color": "rgb(255, 255, 255)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/spell_holy_wordfortitude.gif",
    "name": "Discipline Priest",
    "slug": "discipline-priest",
  },
  {
    "color": "rgb(255, 255, 255)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/spell_holy_guardianspirit.gif",
    "name": "Holy Priest",
    "slug": "holy-priest",
  },
  {
    "color": "rgb(255, 255, 255)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/spell_shadow_shadowwordpain.gif",
    "name": "Shadow Priest",
    "slug": "shadow-priest",
  },
  {
    "color": "rgb(255, 244, 104)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/ability_rogue_eviscerate.gif",
    "name": "Assassination Rogue",
    "slug": "assassination-rogue",
  },
  {
    "color": "rgb(255, 244, 104)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/ability_backstab.gif",
    "name": "Combat Rogue",
    "slug": "combat-rogue",
  },
  {
    "color": "rgb(255, 244, 104)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/ability_stealth.gif",
    "name": "Subtlety Rogue",
    "slug": "subtlety-rogue",
  },
  {
    "color": "rgb(0, 112, 221)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/spell_nature_lightning.gif",
    "name": "Elemental Shaman",
    "slug": "elemental-shaman",
  },
  {
    "color": "rgb(0, 112, 221)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/spell_nature_lightningshield.gif",
    "name": "Enhancement Shaman",
    "slug": "enhancement-shaman",
  },
  {
    "color": "rgb(0, 112, 221)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/spell_nature_magicimmunity.gif",
    "name": "Restoration Shaman",
    "slug": "restoration-shaman",
  },
  {
    "color": "rgb(135, 136, 238)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/spell_shadow_deathcoil.gif",
    "name": "Affliction Warlock",
    "slug": "affliction-warlock",
  },
  {
    "color": "rgb(135, 136, 238)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/spell_shadow_metamorphosis.gif",
    "name": "Demonology Warlock",
    "slug": "demonology-warlock",
  },
  {
    "color": "rgb(135, 136, 238)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/spell_shadow_rainoffire.gif",
    "name": "Destruction Warlock",
    "slug": "destruction-warlock",
  },
  {
    "color": "rgb(198, 155, 109)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/ability_rogue_eviscerate.gif",
    "name": "Arms Warrior",
    "slug": "arms-warrior",
  },
  {
    "color": "rgb(198, 155, 109)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/ability_warrior_innerrage.gif",
    "name": "Fury Warrior",
    "slug": "fury-warrior",
  },
  {
    "color": "rgb(198, 155, 109)",
    "icon": "https://wow.zamimg.com/images/wow/icons/tiny/inv_shield_06.gif",
    "name": "Protection Warrior",
    "slug": "protection-warrior",
  },
]
`;
